### DISCUSSION TITLE
EVP stream cipher random access

### DISCUSSION THREAD

**Discussion by 8ba979b9-7a35-4eed-8a0e-b4dc97eea618** [openssl/openssl#24519] [Q&A] [answered]
hello i have the following question:

is there an officially supported/intended or somewhat algorithm-independent way to specify an offset into the keystream of a streamcipher in the EVP api?

**context**: i am trying to encrypt a file and later read the encrypted file utilizing streamciphers(/modes) such as chacha20 or aes-256-ctr
and during reading allow seeking / random access to that file.
i would like to use the streaming capability for encryption
(initialize EVP_CTX once and encrypt the whole file in one go with few EVP_EncryptUpdates directly)

what i am currently considering:
- for (cipher)block-granularity access:
if you set a starting iv and then call EVP_DecryptUpdate for some bigger (data)block or DecryptUpdate multiple times
then ivs for other starting points would be dependent on how the iv-counter is internally advanced in that (provider-)implementation.
as i understand it, the only alternative is to reset and re-initialize the EVP_CTX for every (cipher)block and provide a manually incremented iv

- for in-block access:
the documentation mentions EVP_CIPHER_CTX_set_num but that does not seem to have the desired effect.
are there perhaps streamcipher-specific OSSL_PARAMs for this purpose?
currently i am decrypting part of a block with garbage data, discard it and then call EVP_DecryptUpdate again with the actual data.

slightly unrelated:
setting the iv similar to this:
```
struct iv {uint64_t something, uint64_t counter};
iv.counter = offset/blocksize;
param_iv = (unsigned char*)iv;
```
would probably lead to problems when trying to decrypt on hardware with different endianness.
does openssl maybe offer some helper functions for such kind of operation?

**Comment by t8m** [comment, marked as answer]
Unfortunately no, there is no algorithm-independent way how to achieve this.

EVP_CIPHER_CTX_set_num() is a legacy concept and won't work with provider based implementations.

However there should be no major performance penalty to run EVP_EncryptInit() with NULL key per some data block (not cipher block) such as 1024 bytes or so with explicitly setting an IV. Doing that would also enable to support AEAD ciphers and protect the integrity of the file as well.

Please note that there are several caveats when designing an encrypted file format and the security properties depend heavily on the design.

**Reply by 8ba979b9-7a35-4eed-8a0e-b4dc97eea618** [reply]
> Unfortunately no, there is no algorithm-independent way how to achieve this.

Ok, thank you for your answer.

> However there should be no major performance penalty to run EVP_EncryptInit() with NULL key per some data block (not cipher block) such as 1024 bytes or so with explicitly setting an IV. Doing that would also enable to support AEAD ciphers and protect the integrity of the file as well.

But do i understand that correctly that EVP_EncryptInit() or probablye EVP_EncryptInit_ex2() too has first to be called with the key and iv and later without resetting the context called again with key=NULL and some other iv?
Are there some requirements calling EncryptInit() again without resetting or can it be called anytime?

**Reply by t8m** [reply]
Yes, you can first call it with the key (and NULL iv) and later with NULL key and the proper iv repeatedly to save cycles. Setting just an iv with the key being kept from the previous invocations should be a fairly fast operation for most ciphers.
